Check BtG key before updating via Capsule Update

mkopec opened the pull request:
Rejects updating via Capsule Update if BtG is enabled and the updated image does not have the same OEM Root Key in the key manifest.

*ref: ncm-2095*

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: Check for BtG key compatibility

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib: Check platform fusing state

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c: remove redundant debug print

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+**/
+STATIC
+BOOLEAN
+IsBootGuardEnabled(
```
```suggestion
IsBootGuardEnabled (
```

Please keep the project coding style.

There are lots of places where function calls miss the space before `(`

mkopec replied:
fb3363242da4043e98652c308c0ab7727de41c80

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+    return FALSE;
+  }
+
+  File = cbfs_get_entry (&Cbfs, "key_manifest.bin");
```
These filenames are configurable in coreboot Kconfig. Would be wise to pass them to EDK2 build system as PCDs, and then get them with `GetPcdPtr`.

Otherwise, we may find out one day, by surprise, that this doesn't work, becasue CBFS filenames are different somewhere else.

Also I thought we talked about empty placeholder manifests. They may be present in the image, which doesn;t necessarily mean a provisioned image. What if placeholder manifests are present? You should at least check if these manifests have a proper signature `__KEYM__` or `__ACBP__`.

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/0908ead1d13b3c535ec986d06ad10f98d461924e for the KM signature check

mkopec replied:
The filenames aren't actually configurable in coreboot's buildsystem

miczyg1 replied:
Ahh sorry. That was probably the ACM binary names that were configurable...

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+  if (File == NULL)
+    return FALSE;
+
+  File = cbfs_get_entry (&Cbfs, "boot_policy_manifest.bin");
```
a/a

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+  KeyManifestHeader *Km = (KeyManifestHeader*) KmBuffer;
+
+  if (Km == NULL ||
+      OemRootKeySize == NULL ||
```
Add check if OemRootKey is NULL, it is both input and output parameter

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/e7b8bf9bc82c192f2e49b8ce5df3d3f04d26e336

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+GetOemRootKeyFromKm (
+  IN CONST VOID   *KmBuffer,
+  IN CONST UINTN  KmSize,
+  OUT VOID        **OemRootKey,
```
```suggestion
  IN OUT VOID        **OemRootKey,
```

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/e7b8bf9bc82c192f2e49b8ce5df3d3f04d26e336

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+  }
+
+  if (!FoundTargetHandle)
+    return FALSE;
```
Is it safe to assume unfused platform if we can't get info about fuse state?

mkopec replied:
Changed to TRUE, let's be careful and assume the platform is fused if we could not determine otherwise: https://github.com/Dasharo/edk2/pull/285/commits/10fe09c0b797c160e8f8e4e874b84c97ba39be83

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+       }
+     } while (!EFI_ERROR(Status));
+
+     if (!FoundTargetHandle) return FALSE;
```
a/a

mkopec replied:
Changed to TRUE, let's be careful and assume the platform is fused if we could not determine otherwise: https://github.com/Dasharo/edk2/pull/285/commits/10fe09c0b797c160e8f8e4e874b84c97ba39be83

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+  RawData = (UINT8 *)SmbiosRecord;
+
+  if (SmbiosRecord->Length < (sizeof(UINT32)))
+    return FALSE;
```
a/a

mkopec replied:
Changed to TRUE, let's be careful and assume the platform is fused if we could not determine otherwise: https://github.com/Dasharo/edk2/pull/285/commits/10fe09c0b797c160e8f8e4e874b84c97ba39be83

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+
+  CopyMem (&Hfsts1, &RawData[0], sizeof(UINT32));
+
+  if ((Hfsts1 & BIT4) == 0)
```
BIT4 is the SPI descriptor lock(ME > 11)/EOM bit (ME11) state, not the fuse state right?

We should check the bit 30 in hfsts6

mkopec replied:
ad8604f8c0ae5a8a7dcdd794ed2d15e2d3971295

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.h:
```diff
@@ -46,4 +46,24 @@ IsRangeWriteable (
+  IN CONST UINTN ImageSize
+  );
+
+#endif // FLASHING_H__
```
Fix new line please

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/0339bc0879b7d10e025d81012f0123c045826f6b

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -1015,6 +1016,17 @@ FmpDeviceSetImageWithStatus (
 
+  if (!AreImageBtgKeysCompatible(CurrentImage, Image, ImageSize)) {
+    FreePool (CurrentImage);
+    AsciiPrint("New image is not signed with a compatible OEM Root Key, refusing to flash\n");
```
```suggestion
    AsciiPrint("New image is not signed with a compatible OEM Root Key, aborting update\n");

```

miczyg1 replied:
Sounds less agressive I think

mkopec replied:
d35d3a6c88263639d9de327cda2f3f9b6cd28f3f

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+  VOID               *CurrentOemKey, *UpdatedOemKey;
+  UINTN              CurrentOemKeyLen, UpdatedOemKeyLen;
+
+  // Platform is unfused, so different BtG key doesn't matter.
```
That's only true if ME region is not locked. Shouldn't we check if ME is unlocked to flash an image with different keys?

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/5e0cbbb5485e7bd6ffc69d8b88336188423315ec

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -1015,6 +1016,17 @@ FmpDeviceSetImageWithStatus (
+    AsciiPrint("New image is not signed with a compatible OEM Root Key, refusing to flash\n");
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a(): New image is not signed with a compatible OEM Root Key, refusing to flash\n",
```
```suggestion
      "%a(): New image is not signed with a compatible OEM Root Key, aborting update\n",

```

mkopec replied:
d35d3a6c88263639d9de327cda2f3f9b6cd28f3f

miczyg1 commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+  if (Km == NULL ||
+      OemRootKeySize == NULL ||
+      KmSize < sizeof(KeyManifestHeader) ||
+      KmSize < Km->KeySignatureOffset + sizeof(KeyAndSigHeader) + sizeof(BtgPubKey))
```
Maybe add a check for KM signature: `__KEYM__` if the manifest is even valid. Because KeySignatureOffset may be 0, so this condition may be satisfied when a placeholder KM is used.

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/0908ead1d13b3c535ec986d06ad10f98d461924e

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +835,198 @@ MergeFirmwareImages (
+  *OemRootKeySize = sizeof(BtgPubKey) + PubKeyOffset->KeySize / 8;
+
+  PubKey = AllocatePool (*OemRootKeySize);
+  CopyMem(PubKey, PubKeyOffset, *OemRootKeySize);
```
There is a convenient shortcut for making a copy: `PubKey = AllocateCopyPool(PubKeyOffset, *OemRootKeySize);`.  However, why use dynamic allocation at all?  You're only reading the data inside of images while they exist.  As is, the code leaks the memory allocated here.

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/3b5485a046b524fefcda29b2f758bd66fe44a493

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +835,198 @@ MergeFirmwareImages (
+    return EFI_INVALID_PARAMETER;
+
+  if (PubKeyOffset->Exponent != 0x10001)
+    return EFI_INVALID_PARAMETER;
```
Are these checks done for validation that this is in fact a key?  Checking fields of `KeyAndSigHeader` should avoid processing keys of unhandled types (e.g., ECC) and not limit key size or demand a specific exponent (even though it will probably be `0x10001`).

mkopec replied:
Yes these checks are for validation if this is indeed a key. Changed to checking the key type instead (RSA): https://github.com/Dasharo/edk2/pull/285/commits/be5faada8f25f1ff84a9b9d3cda0c5c6483dac15

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -930,6 +933,126 @@ GetOemRootKeyFromKm (
+
+  Status = gBS->LocateProtocol (&gEfiSmbiosProtocolGuid, NULL, (VOID **)&Smbios);
+  if (EFI_ERROR (Status))
+    return FALSE;
```
Unfused if no SMBIOS protocol is found?  This shouldn't even happen, but sounds somewhat weird.

mkopec replied:
Changed to TRUE, let's be careful and assume the platform is fused if we could not determine otherwise: https://github.com/Dasharo/edk2/pull/285/commits/10fe09c0b797c160e8f8e4e874b84c97ba39be83

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+    TempStruct.Hdr = SmbiosRecord;
+
+    GroupName = GetSmbiosString (TempStruct, Type14->GroupName);
+    if (GroupName != NULL && AsciiStrCmp (GroupName, "$MEI") == 0) {
```
Do we want to check that `Type14->ItemType` is `0xdb` (arbitrary value set in coreboot?)?

mkopec replied:
Not sure. @miczyg1 mentioned this could be different on other BIOS implementations but that doesn't matter much here.

miczyg1 replied:
> Do we want to check that Type14->ItemType is 0xdb (arbitrary value set in coreboot?)?

We hardcode that to 0xdb, but the type is BIOS-build-dependent, when it comes to proprietary BIOS. Also structrues of given type may occur more than once. `Handle` is what uniquely identifies the entry we search for.

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +838,320 @@ MergeFirmwareImages (
+  if (SmbiosRecord->Length < (sizeof(UINT32)))
+    return FALSE;
+
+  CopyMem (&Hfsts1, &RawData[0], sizeof(UINT32));
```
This seems wrong.  `RawData` points at the start of the record, not its data.  Also, the whole record looks like this:
```c
struct fwsts_record {
	u8 heci_name;
	u32 reg[6];
} __packed;

struct fwsts_smbios_table {
	struct smbios_header header;
	u8 version;
	u8 count;
	struct fwsts_record record[CONFIG_MAX_MEI_DEVICES];
	u8 eos[2];
} __packed;
```
So need to reach into `record[0].reg[5]` (HFSTS6; always first HECI?).  Currently the header seems to be copied.

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/88b3336bdd555a31fa6fcd7ee633295c4c2bca65

SergiiDmytruk reviewed (changes requested):
Came back after looking at coreboot changes.

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: fixes for code style

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.h: add missing newline

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: check EOM at HFSTS6 bit 30

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c: reword BtG key fail message

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: fix build error

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: add checks for OemRootKey

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib: Move IsDescriptorLocked to Flashing.c

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib: Allow BtG key change only if descriptor is unlocked

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: Properly cast FWSTS record table

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: remove redundant memory alloc & copy

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: indentation fix

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: Check KM signature

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: Check BtG key type

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: fix build error

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: separete if conditions with parentheses

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c: assume BTG is fused if could not determine from SMBIOS

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

SergiiDmytruk reviewed (dismissed):
Needs testing, but looks good now.  coreboot PR will need to have EDK hash updated.

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib: Show a popup when BtG keys don't match FPF

Show a 2x scaled popup on the screen to inform the user what's happened
and what to do in this situation. Print some debug info to the screen to
provide to developers.

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -992,12 +1371,20 @@ FmpDeviceSetImageWithStatus (
+
+    ShowBtgErrorPopup(CurrentImage, ImageSize, ErrorString);
+
+    FreePool(ErrorString);
```
No need to free a string literal.

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/170d6b39ccff98e61582de6a7547bc012bdce27a

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -861,6 +865,380 @@ QueryVariableInfoHook (
+        L"Press ENTER to boot normally."
+        );
+
+    Status = gBS->WaitForEvent (1, Events, &Index);
```
`Events` should also include `TimerEvent`.  As is, this `while` loop will likely never stop executing.

mkopec replied:
It should actually never stop unless enter is explicitly pressed. We want the user to confirm they've read the message. Will clean up

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/4a1c766b7e6f312305409a6717ea25d003ebd6bd

SergiiDmytruk replied:
The code still creates a timer event.

mkopec replied:
whoops sorry: https://github.com/Dasharo/edk2/pull/285/commits/8d3ee0305e4dde96d4a4656479f3c33476924e19

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -9,10 +9,12 @@
 #include <Guid/SystemResourceTable.h>
 #include <Library/BaseMemoryLib.h>
 #include <Library/BlParseLib.h>
+#include <Library/BootLogoLib.h>
```
I think this library is not actually used but gets added here and in the `.inf`-file.

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/9bb64f060082cd5cc5ee86e248213242320127a6

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -861,6 +865,380 @@ QueryVariableInfoHook (
+// Configuration
+#define SCALE_FACTOR        2
+#define GLYPH_WIDTH         8
+#define GLYPH_HEIGHT        19
```
Glyph dimensions are fixed by the font?

mkopec replied:
It's fixed in the font and also defined like this in a couple of places in the edk2 tree

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -861,6 +865,380 @@ QueryVariableInfoHook (
+}
+
+// Configuration
+#define SCALE_FACTOR        2
```
Should this be determined by looking at resolution?  If it's HiDPI, then scale by a factor of 2 and don't otherwise.  I'm just unsure we'll notice an enormous dialog during testing for some other devices unless we'll test this case.

mkopec replied:
Made scale factor adjustable: https://github.com/Dasharo/edk2/pull/285/commits/4b07561bb88efe003204a25551871e14217ea1a3

mkopec replied:
and made it so that if HRes>1920 only then do we scale 2x: https://github.com/Dasharo/edk2/pull/285/commits/fae3a925a4aaa3134049a6dc1c70b8d98b74e05d

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -861,6 +865,380 @@ QueryVariableInfoHook (
+  UINTN CurrentY = BoxY + (PADDING_Y * SCALE_FACTOR);
+  UINTN TextStartX = BoxX + (PADDING_X * SCALE_FACTOR);
+
+  VA_START(Args, NumberOfLines);
```
Can't do this twice (it may work, but it's not guaranteed).  Should use `VA_COPY` after the first `VA_START` and before `VA_END`.

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/8c7a3ef7f3688feb5d410ae92573b36403c15146

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -1099,6 +1100,100 @@ IsPlatformFused (
+  HashInput = AllocatePool (HashInputLen);
+
+  CopyMem (HashInput, OemKey->Modulus, OemKey->KeySize / 8);
+  CopyMem (HashInput + (OemKey->KeySize / 8), &OemKey->Exponent, sizeof(OemKey->Exponent));
```
Were modulus and exponent swapped for some cryptographic reason?

mkopec replied:
That's just how Intel does it, I don't know what the justification for this is. The hash calculated here matches what's shown in Intel MFIT in the OEM Key Hash field.

SergiiDmytruk replied:
Then a comment saying that ordering of fields and digest algorithm matches behaviour of Intel's MFIT could be helpful.  The fact of hashing looked weird as well, but I thought it's done to shorten warning message.

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/edf5f1a73ab14232c7e38ec30857bffe6b16d586

The key is basically hashed to reduce the number of E-fuses needed for Boot Guard, as I understand it. The hash is what is actually burnt into FPFs.

SergiiDmytruk reviewed (changes requested):
A few things to fix.  Haven't tested the dialog yet, will also need to go through its implementation in more detail.

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib: remove unused lib

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c: don't free string literal

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c: copy variadic arguments with va_copy

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c: remove 10s timer remnants

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c: Parametrize popup scaling factor

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c: Adjust scale factor based on resolution

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c: remove timer event

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.*: explain why OEM key is hashed

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec commented:
Looks like this: 
<img width="1920" height="1440" alt="image" src="https://github.com/user-attachments/assets/14e3b561-5a99-42db-b8b4-956d54d7e8c5" />

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -861,29 +865,392 @@ QueryVariableInfoHook (
+    ClearScreen();
+    CreateMultiStringPopUpScaled (
+        GetDisplayScaleFactor(),
+        100,
```
```suggestion
        96,
```
With `100`:
```
ScreenWidth=800
ScreenHeight=600
RequestedWidth=100
UnscaledContentW=800
UnscaledContentH=247
BoxW=832
BoxH=285
```

<img width="1600" height="1248" alt="Image" src="https://github.com/user-attachments/assets/a15d9add-b23f-4dae-a886-3b66848ec82b" />

While `96` fits perfectly:
```
RequestedWidth=96
ScreenWidth=800
ScreenHeight=600
RequestedWidth=96
UnscaledContentW=768
UnscaledContentH=247
BoxW=800
BoxH=285
```

<img width="1600" height="1248" alt="Image" src="https://github.com/user-attachments/assets/cf0feae1-9c2f-4fc1-a27a-1625a092967c" />

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/b3f0caeafcb3cde4d3bf8d39d1cfae12a7c9fb10

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -861,29 +865,392 @@ QueryVariableInfoHook (
+  // We allocate enough space for one line at a time
+  UINTN LinePixels1x = UnscaledContentW * GLYPH_HEIGHT;
+  RenderBuffer1x = AllocateZeroPool(LinePixels1x * sizeof(EFI_GRAPHICS_OUTPUT_BLT_PIXEL));
+  RenderBufferScaled = AllocateZeroPool(LinePixels1x * ScaleFactor * ScaleFactor * sizeof(EFI_GRAPHICS_OUTPUT_BLT_PIXEL));
```
Something seems to be trashing memory:
```
DXE_ASSERT!: [DxeCore] .../coreboot/payloads/external/edk2/workspace/Dasharo/MdeModulePkg/Core/Dxe/Mem/Pool.c (741): Tail->Signature == ((('p') | ('t' << 8)) | ((('a') | ('l' << 8)) << 16))

DXE_ASSERT!: [DxeCore] .../coreboot/payloads/external/edk2/workspace/Dasharo/MdeModulePkg/Core/Dxe/Mem/Pool.c (742): Head->Size == Tail->Size

ASSERT_EFI_ERROR (Status = Invalid Parameter)

DXE_ASSERT!: [FmpDxe] .../coreboot/payloads/external/edk2/workspace/Dasharo/MdePkg/Library/UefiMemoryAllocationLib/MemoryAllocationLib.c (821): !(((RETURN_STATUS)(Status)) >= 0x8000000000000000ULL)
```

If I allocate 10 times the amount on this line the errors are gone, but I don't see this buffer being overrun in the code.  Maybe the issue in a different place.

mkopec replied:
I'm not sure what to do here, how to address this, @SergiiDmytruk do you have suggestions?

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c: shrink popup to fit 800x600

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -861,29 +865,392 @@ QueryVariableInfoHook (
+      // Fill the scaled block
+      for (UINTN Index = 0; Index < ScaleFactor; ++Index) {
+        Dest[DestIndex + Index]             = Pixel;
+        Dest[DestIndex + DestWidth + Index] = Pixel;
```
Here is the cause of memory error.  This assumes scale factor of 2.  Need to add another loop over `ScaleFactor` to update appropriate number of lines in the destination.

mkopec replied:
https://github.com/Dasharo/edk2/pull/285/commits/de7e555a0786ee4ebac0b92d867fca8f69c9696d

mkopec committed:
DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c: fix scaling calculations

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

SergiiDmytruk commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c:
```diff
@@ -797,3 +860,394 @@ MergeFirmwareImages (
+  if (!IsBootGuardEnabled (Current, ImageSize))
+    return TRUE;
+
+  // Going from BtG -> No BtG is not possible if the platform is fused.
```
The comment is for an older version of the code which checked for fused state but not descriptor's lock state.  Now we can get here when:
- platform is fused, descriptor is locked (keys must match)
- platform is fused, descriptor is unlocked (won't happen with current configuration)
- platform is unfused, descriptor is locked (do we need to check BootGuard keys in this case?  is this even possible/relevant?)

mkopec replied:
It is relevant if the descriptor is locked, because if the ME is provisioned for some bootguard keys, BtG won't accept an unsigned BIOS region or one signed with other keys. BtG will prevent the platform from booting even though the fuses aren't blown.

If the descriptor was unlocked, we'd just update the ME to match the BIOS and everything would boot fine.

SergiiDmytruk reviewed (approved):
LGTM, just one question to be sure we've thought about all cases here.
